supabase/functions/_script/backup_and_cleanup.ts

```diff
@@ -0,0 +1,98 @@
+import { createClient } from 'https://esm.sh/@supabase/supabase-js@^2.2.3';
+import 'https://deno.land/x/dotenv/load.ts';
+import { r2 } from '../_utils/r2';
+import type { Database } from '../_utils/supabase.types.ts';
+import * as fs from 'https://deno.land/std/fs/mod.ts';
+import * as path from 'https://deno.land/std/path/mod.ts';
+
+const supabaseUrl = Deno.env.get('SUPABASE_URL');
+const supabaseServiceRoleKey = Deno.env.get('SUPABASE_SERVICE_ROLE_KEY');
+const backupFolder = 'backup';
+
+if (!supabaseUrl || !supabaseServiceRoleKey) {
+  console.error('Error: SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY environment variables are required.');
+  Deno.exit(1);
+}
+
+function useSupabase() {
+  const options = {
+    auth: {
+      autoRefreshToken: true,
+      persistSession: true,
+      detectSessionInUrl: false,
+    },
+  };
+  return createClient<Database>(supabaseUrl, supabaseServiceRoleKey, options);
+}
+
+async function backupAndDeleteOldEntries() {
+  try {
+    if (!await fs.exists(backupFolder)) {
+      await fs.ensureDir(backupFolder);
+    }
+
+    const supabase = useSupabase();
+    const daysToKeep = 30;
+    const pageSize = 1000;
+
+    const currentDate = new Date();
+    const cutoffDate = new Date(currentDate.getTime() - daysToKeep * 24 * 60 * 60 * 1000);
+
+    let offset = 0;
+    let hasMoreData = true;
+    let fileNumber = 0;
+    let recordsToDelete: Database['public']['Tables']['app_versions']['Row'][] = []; 
+
+    while (hasMoreData) {
+      const { data: oldEntries, error: queryError } = await supabase
+        .from<Database['public']['Tables']['app_versions']['Row']>('app_versions')
+        .select()
+        .lt('updated_at', cutoffDate.toISOString())
+        .range(offset, offset + pageSize - 1);
+
+      if (queryError) {
+        console.error('Error querying the Supabase table:', queryError.message);
+        break;
+      }
+
+      if (!oldEntries || oldEntries.length === 0) {
+        hasMoreData = false;
+        break;
+      }
+
+      const backupData = JSON.stringify(oldEntries);
+      const backupFilename = `backup-${currentDate.toISOString()}-page-${offset}-${offset + pageSize - 1}.json`;
+      const backupPath = path.join(backupFolder, backupFilename);
+
+      try {
+        await r2.upload(backupPath, new TextEncoder().encode(backupData));
+        console.log(`Backup saved to R2: ${backupPath}`);
+        recordsToDelete = recordsToDelete.concat(oldEntries);
+      } catch (backupError) {
+        console.error('Error saving backup to R2:', backupError);
+        throw new Error('Backup to R2 failed. Stopping execution to prevent data loss.');
+      }
```

Comment:
I think the last thing to optimize is that the case:

- The loop ran fine for the first iteration, now recordsToDelete has some X records to be deleted.
- The loop errored for the second iteration, which now throws a new Error in catch and terminates the whole script.
- The recordsToDelete that's outside the loop set to be deleted won't fire, and while the original X items are backed up, aren't deleted from the supabase.
- When the script runs again, they are backed up again leading to duplicate calls.

Reply:
@ayusham001 can you close this conversation as well?